File: Algorithms/Graph/bfs-dfs.cpp
#include <bits/stdc++.h>
#include <iostream>
#include <queue>
#include <set>
#include <utility>
#include <vector>

using namespace std;

map<string, vector<string>> graph;

void depth_first_search(string source,
                        set<string> *visited = new set<string>()) {
  // Print
  cout << source << " ";

  // Node visited
  visited->insert(source);

  // Doesn't have neighbours
  if (!graph.count(source))
    return;

  // Meet neighbours
  for (auto v : graph[source])
    depth_first_search(v, visited); // Recurse
}

void breadth_first_search(string source) {
  set<string> visited = {};

  // Make a queue
  queue<string> queue;
  queue.push(source);

  while (!queue.empty()) {
    string u = queue.front();
    queue.pop();

    cout << u << " ";
    visited.insert(u);

    if (!graph.count(u))
      continue;

    for (string w : graph[u])
      if (!visited.count(w))
        queue.push(w);
  }
}

int main() {
  graph = {
      {"0", {"1", "2"}},
      {"1", {"3"}},
      {"3", {"4", "5"}},
      {"2", {"6", "7"}},
      {"6", {"8", "9"}},
  };

  cout << "Depth First Search : ";
  depth_first_search("0");
  cout << endl;

  cout << "Breadth First Search : ";
  breadth_first_search("0");
  cout << endl;

  return 0;
}

File: Algorithms/Graph/bfs-dfs.go
package main

import "fmt"

var graph map[string][]string

func depth_first_search(source string, visited map[string]bool) {
	fmt.Print(source, " ")
	visited[source] = true

	if _, found := graph[source]; !found {
		return
	}

	for _, w := range graph[source] {
		// fmt.Print(value, " ")
		if !visited[w] {
			depth_first_search(w, visited)
		}
	}
}

func breadth_first_search(source string) {
	queue := make([]string, 0)
	visited := make(map[string]bool)

	queue = append(queue, source)

	for len(queue) > 0 {
		// Get front of queue
		u := queue[0]

		// Print element
		fmt.Print(u, " ")

		// Remove front of queue
		queue = queue[1:]

		// Mark node as visited
		visited[u] = true

		// Check if node exists
		if _, found := graph[u]; !found {
			continue
		}

		for _, w := range graph[u] {
			if !visited[w] {
				queue = append(queue, w)
			}
		}
	}
}

func main() {
	graph = map[string][]string{
		"0": {"1", "2"},
		"1": {"3"},
		"3": {"4", "5"},
		"2": {"6", "7"},
		"6": {"8", "9"},
	}

	fmt.Print("Depth First Search : ")
	depth_first_search("0", map[string]bool{})
	fmt.Println()

	fmt.Print("Breadth First Search : ")
	breadth_first_search("0")
	fmt.Println()
}
